### ISSUE TITLE
Accessing git in parallel hooks

### ISSUE THREAD

**Issue by saveman71** [dirk/quickhook#20]
Thanks for the great, straightforward tool! After the last upgrade a colleague has been seeing the following:

```
check-conflicts: git is not allowed in parallel hooks (git diff --cached --name-only -G<<<<<|=====|>>>>>)
```

In our hooks we're accessing git in a way that I believe can be parallel (I may be wrong):

```bash
#!/bin/bash

# Simple check for merge conflics
conflicts=`git diff --cached --name-only -G"<<<<<|=====|>>>>>"`


# Something went wrong
if [[ -n "$conflicts" ]]; then
    echo
    echo "Unresolved merge conflicts in these files:"

    for conflict in $conflicts; do
        echo $conflict
    done;

    exit 1;
fi

exit 0
```

or

```bash
#!/bin/bash

# Checks the the staged JSON files are valid JSON
#
# There may be a better way to do this, but for now it works by calling
# 'python -m json.tool' on the file, which will exit with a non-zero status if
# the json is invalid:
#
# $ echo "{" | python -m json.tool
# Expecting property name enclosed in double quotes: line 2 column 1 (char 2)

files=$(git diff --cached --name-status | awk '$1 != "D" && $NF ~ /\.json$/ { print $NF }')

for file in $files; do
  echo "$file"
  git show :$file $file | python -m json.tool > /dev/null || exit 1
done;
```

It's always a combination of git diff or git show. Are those really risky to use in parallel hooks?

(weirdly only the check-conflicts hook warns)

**Comment by dirk** [comment]
Hey @saveman71, sorry that I missed the notification and didn't respond sooner, but thank you for the feedback! 🤦‍♂️ Would moving these hooks into the [`pre-commit-mutating` directory](https://github.com/dirk/quickhook#mutating-hooks) be a viable option for y'all? Those hooks run sequentially—so you might loose a bit of performance—but on the upside Quickhook doesn't put a Git safety shim in their `PATH`.

You're probably right that `git diff` and `git show` can be accessed in parallel and are safe enough to do so. I'll dig in tonight, and if they are then I'll improve the shim to safelist them and forward the call to the real Git. 🤞

**Comment by dirk** [comment]
Cut a new [v1.6.1 release](https://github.com/dirk/quickhook/releases/tag/v1.6.1) with the shim allowing a lot of the common read-only commands through to `git`. The macOS Homebrew tap is also updated: dirk/homebrew-quickhook@659e55fba50603d21daed84074d987919c55e560. Hope that helps!

**Comment by saveman71** [comment]
Hi @dirk, same thing, didn't get a notification :facepalm: Will test!

So after testing a bit, I have had two problems:

1. I use `git grep` which isn't in your list. It seems quite thread safe / read-only as far as I can see from my 2 minutes lookup online, lmk if I should open a PR to add it
2. I work in the Elixir world so I use [Mix](https://elixir-lang.org/getting-started/mix-otp/introduction-to-mix.html) -- Elixir's package manager. It can handle dependencies with git, and you invoke scripts via mix. There's also `mix format` which I run in a pre-commit hook. It tries to lookup if the dependency tree is up to date with lockfiles and seems to fail because the git commands it uses are also shimmed.
  Here's the file listing the command it uses: https://github.com/elixir-lang/elixir/blob/main/lib/mix/lib/mix/scm/git.ex
  I can see:
	a. `git version`
	b. `git remote`
    c. `git config` => that definitely writes some files so not parallel safe :/
  For now I added some flags to tell mix not to check dependencies so I consider this as fixed.

Thanks again, it's nice to have the command outputs better formatted, as I was using 1.4 until now

**Comment by dirk** [comment]
Added `git grep` to the allowed list in a6aa851821c3822b29104486799158fee920c084, released this morning in v1.6.2! The [`pre-commit-mutating` hooks](https://github.com/dirk/quickhook#mutating-hooks) should work for having Mix check dependencies: Git is not shimmed on the $PATH of those hooks, as they're run sequentially and therefore can't conflict with other hooks.

**Comment by saveman71** [comment]
Thanks for git grep @dirk ! As for mix since these are long tasks (can be up to a few seconds), it's best if I can keep them // but the work arounds works perfectly.
